### ISSUE TITLE
An entry in the Proxy.proxies storage map cannot be decoded

### ISSUE THREAD

**Issue by jsdw** [polkadot-fellows/runtimes#453]
Hello!

While working on decoding historic storage entries, I ran into an issue whereby the storage entry in `Proxy.Proxies` keyed by account ID (twox64_concated) `0x0E6DE68B13B82479FBE988AB9ECB16BAD446B67B993CDD9198CD41C7C6259C49` and with value `0x0898c8a3d01da9877b7b30877c717ae8f2a7e726cefa176c5dfcdcebc9b6a122230098c8a3d01da9877b7b30877c717ae8f2a7e726cefa176c5dfcdcebc9b6a1222304005109bd2e0000000000000000000000` fails to decode. it decodes OK in spec version 18, and starts failing to decode as of spec version 23 and until the present day by the looks of it.

The value type for this map in spec version 18 is `(Vec<(T::AccountId, T::ProxyType)>, BalanceOf<T>)`.

The value type as of spec version 23 is `(Vec<ProxyDefinition<T::AccountId, T::ProxyType, T::BlockNumber>>, BalanceOf<T>)`, where `ProxyDefinition` is:

```rust
#[derive(Encode, Decode, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug)]
pub struct ProxyDefinition<AccountId, ProxyType, BlockNumber> {
	/// The account which may act on behalf of another.
	delegate: AccountId,
	/// A value defining the subset of calls that it is allowed to make.
	proxy_type: ProxyType,
	/// The number of blocks that an announcement must be in place for before the corresponding call
	/// may be dispatched. If zero, then no announcement is needed.
	delay: BlockNumber,
}
```

In other words, the size of the value type grew by 4 bytes when this `delay: BlockNumber` field was added (and then the tuple turned to a struct).

In my tests, the value of this field never changed when the type did, and so the value is unable to be decoded to this day.

Should a storage migration be applied to remove this entry from the map?

**Comment by jsdw** [comment]
Kian also pointed me to https://github.com/paritytech/polkadot-sdk/pull/1805/files and https://paritytech.github.io/polkadot-sdk/master/asset_hub_westend_runtime/struct.DeleteUndecodableStorage.html, which may help.

One thought in this case was that the storage entry in Proxy.proxies that fails _does_ in fact decode without errors, but it leaves some undecoded bytes at the end that weren't decoded. If we have a CI check in place to test that storage can be decoded, it'd be worth making sure that it also checks that all bytes are consumed during decoding.

**Comment by kianenigma** [comment]
The lack of `env.CHECKS` is the main issue: https://github.com/polkadot-fellows/runtimes/blob/e6b9132701df4311e7d6331e3a9e051ae8de0104/.github/workflows/check-migrations.yml#L93

@ggwpez / @mordamax I am guessing this was a delibrate choice to save us CI time right? or else do you know why we don't pass this arg in?

**Comment by ggwpez** [comment]
The relay decode checks are broken, i think because of nompools or something?  
I had to disable them here: https://github.com/polkadot-fellows/runtimes/blob/e6b9132701df4311e7d6331e3a9e051ae8de0104/.github/workflows/check-migrations.yml#L64-L65

But this disables them for all pallets...

**Comment by Stephenlawrence00** [comment]
@kianenigma @ggwpez opened #1207 to fix this, a migration that repairs the `Proxy.Proxies` entries left undecodable since spec 23 (keeping valid proxies and refunding deposits where none survive), would you mind taking a look please?
